src/providers/google-vertex-ai/chatComplete.ts

```diff
@@ -0,0 +1,380 @@
+export const GoogleChatCompleteResponseTransform: (
+  response:
+    | GoogleGenerateContentResponse
+    | GoogleErrorResponse
+    | GoogleErrorResponse[],
+  responseStatus: number
+) => ChatCompletionResponse | ErrorResponse = (response, responseStatus) => {
+  // when error occurs on streaming request, the response is an array of errors.
+  if (
+    responseStatus !== 200 &&
+    Array.isArray(response) &&
+    response.length > 0 &&
+    'error' in response[0]
+  ) {
+    const { error } = response[0];
+    return {
+      error: {
+        message: error.message,
+        type: error.status,
+        param: null,
+        code: String(error.code),
+      },
+      provider: GOOGLE_VERTEX_AI,
+    } as ErrorResponse;
+  }
+
+  if (responseStatus !== 200 && 'error' in response) {
+    return {
+      error: {
+        message: response.error.message ?? '',
+        type: response.error.status ?? null,
+        param: null,
+        code: String(response.error.code),
+      },
+      provider: GOOGLE_VERTEX_AI,
+    } as ErrorResponse;
+  }
+
+  if ('candidates' in response) {
+    const {
+      promptTokenCount = 0,
+      candidatesTokenCount = 0,
+      totalTokenCount = 0,
+    } = response.usageMetadata;
+
+    return {
+      id: crypto.randomUUID(),
+      object: 'chat_completion',
+      created: Math.floor(Date.now() / 1000),
+      model: 'Unknown',
+      provider: GOOGLE_VERTEX_AI,
+      choices:
+        response.candidates?.map((generation, index) => {
+          let message: Message = { role: 'assistant', content: '' };
+          if (generation.content.parts[0]?.text) {
+            message = {
+              role: 'assistant',
+              content: generation.content.parts[0]?.text,
+            };
+          } else if (generation.content.parts[0]?.functionCall) {
+            message = {
+              role: 'assistant',
+              tool_calls: [
+                {
+                  id: crypto.randomUUID(),
+                  type: 'function',
+                  function: {
+                    name: generation.content.parts[0]?.functionCall.name,
+                    arguments: JSON.stringify(
+                      generation.content.parts[0]?.functionCall.args
+                    ),
+                  },
+                },
+              ],
+            };
+          }
+          return {
+            message: message,
+            index: index,
+            finish_reason: generation.finishReason,
+          };
+        }) ?? [],
+      usage: {
+        prompt_tokens: promptTokenCount,
+        completion_tokens: candidatesTokenCount,
+        total_tokens: totalTokenCount,
+      },
+    };
+  }
+
+  return {
+    error: {
+      message: `Invalid response recieved from google: ${JSON.stringify(
+        response
+      )}`,
+      type: null,
+      param: null,
+      code: null,
+    },
+    provider: GOOGLE_VERTEX_AI,
+  } as ErrorResponse;
+};
+
+export const GoogleChatCompleteStreamChunkTransform: (
```

Comment:
The last chunk of the stream sends usageMetadata object which has the token usage. We can do the same for transformed stream. The last chunk can have usage object which is in OpenAI format. Checkout bedrock model transformers for reference.

This is the object that vertex sends in last chunk:
```
"usageMetadata": {
        "promptTokenCount": 8,
        "candidatesTokenCount": 181,
        "totalTokenCount": 189
    }
    ```